### PR TITLE
Update tinydb example

### PR DESCRIPTION
**Pull request by JohnDoneth**
I've updated the tinydb example to use the new Async-Await feature set and to use `await` wherever possible.

## Solution

I've changed the typical combinator method on listeners from 
```rust
listener
        .incoming()
        .map_err(|e| println!("error accepting socket; error = {:?}", e))
        .for_each(move |socket| { ... }
```
to the new async pattern of `await`'ing on each new connection.
```rust
match listener.accept().await {
    Ok((socket, _)) => { ... }
```
- The logic for the connection handling is in an `async` closure and passed to `tokio::spawn`
- Requests are now decoded and encoded using `Framed` using `LinesCodec`.

### CODE REVIEW

File: tokio/examples/tinydb.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+#![feature(async_closure)]
+#![deny(warnings, rust_2018_idioms)]
+
+use std::collections::HashMap;
+use std::env;
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+
+use tokio;
+use tokio::codec::Framed;
```
Comment (LucioFranco) [inline tokio/examples/tinydb.rs]: small nit but can we combine these imports, so `tokio::codec::{Framed, LinesCodec}`

File: tokio/examples/tinydb.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+use tokio::codec::LinesCodec;
+use tokio::net::TcpListener;
+
+use futures::SinkExt;
```
Comment (LucioFranco) [inline tokio/examples/tinydb.rs]: same here

File: tokio/examples/tinydb.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    let db = Arc::new(Database {
+        map: Mutex::new(initial_db),
+    });
+
+    loop {
+        match listener.accept().await {
+            Ok((socket, _)) => {
+                // Here's where the meat of the processing in this server happens. First
+                // we see a clone of the database being created, which is creating a
+                // new reference for this connected client to use. Also note the `move`
```
Comment (LucioFranco) [inline tokio/examples/tinydb.rs]: this should be updated move isnt needed.

Comment (JohnDoneth) [reply tokio/examples/tinydb.rs]: Without `move` there the compiler errors with "`db` doesn't live long enough".

Comment (LucioFranco) [reply tokio/examples/tinydb.rs]: Sorry what I mean this comment I believe is referring to `listener.incoming().for_each(move || {..})` which was the old way to do this but now that we have async/await db is captured by ref here.

File: tokio/examples/tinydb.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+//! Namely you can issue two forms of commands:
+//!
+//! * `GET $key` - this will fetch the value of `$key` from the database and
+//!   return it. The server's database is initially populated with the key `foo`
+//!   set to the value `bar`
+//! * `SET $key $value` - this will set the value of `$key` to `$value`,
+//!   returning the previous value, if any.
+
+#![feature(async_await)]
+#![feature(async_closure)]
```
Comment (LucioFranco) [inline tokio/examples/tinydb.rs]: this feature should be dropped since I don't think its being used and won't be stabilized with async_await.

Comment (JohnDoneth) [reply tokio/examples/tinydb.rs]: That's strange, I wonder how that got in there. The rest works fine without it. Good catch!

**Comment by LucioFranco** [comment]
Will merge once CI passes.

**Comment by carllerche** [comment]
Thanks all for getting this done :+1:

**Pull request merged**